jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,395 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * A fast container of poolable objects, with optional support for
+ * multiplexing and thread-local caching.
+ * @param <T>
+ */
+public class Pool<T> implements AutoCloseable
+{
+    private static final Logger LOGGER = Log.getLogger(Pool.class);
+
+    private final List<Entry> sharedList = new CopyOnWriteArrayList<>();
+    private final Collection<List<Entry>> allCaches = new CopyOnWriteArrayList<>();
+    private final ThreadLocal<List<Entry>> cache;
+    private final Lock closeLock = new ReentrantLock();
+    private final int cacheSize;
+    private volatile boolean closed;
+    private volatile int maxMultiplex = 1;
+    private volatile int maxUsageCount = -1;
+
+    /**
+     * Construct a Pool with the specified thread-local cache size.
+     *
+     * @param cacheSize the thread-local cache size. A value of 0 means the cache is disabled.
+     */
+    public Pool(int cacheSize)
+    {
+        if (cacheSize < 0)
+            throw new IllegalArgumentException("Cache size must be >= 1");
```

Comment:
`cacheSize` can be `0`. Perhaps just remove the throw and assume `<= 0` means disabled.